### ISSUE TITLE
postgresql 9.3 fails to install on ubuntu 14.04 with non default data_directory

### ISSUE THREAD

**Issue by phcyso** [hw-cookbooks/postgresql#248]
Hi,

We ran into a problem when trying to install postgres 9.3 on ubuntu 14.04 using version 3.4.18 of the cookbook.
The main issue is the default database files need to be in place to start the server and the data_directory doesn't get created by anything in the recipe.

To fix this internally we have modified the server_debian.rb recipe to contain the following: 

```
include_recipe "postgresql::client"

node['postgresql']['server']['packages'].each do |pg_pack|

  package pg_pack

end

#Create the data directory if it doesnt exist.
directory node['postgresql']['config']['data_directory'] do
  mode '0700'
  owner 'postgres'
  group 'postgres'
  recursive true
  notifies :stop, 'service[postgresql]', :immediately
end

execute "create base database in data_directory" do
	user 'postgres'
	command "PATH=$PATH:/usr/lib/postgresql/#{node['postgresql']['version']}/bin initdb  #{node['postgresql']['config']['data_directory']}"
  	not_if { ::FileTest.exist?(File.join(node['postgresql']['config']['data_directory'], "PG_VERSION")) }
end

execute 'Set locale and Create cluster' do
  command 'export LC_ALL=C; /usr/bin/pg_createcluster --start ' + node['postgresql']['version'] + ' main'
  action :run
  not_if { ::File.directory?('/etc/postgresql/' + node['postgresql']['version'] + '/main') }
end
include_recipe "postgresql::server_conf"

service "postgresql" do
  service_name node['postgresql']['server']['service_name']
  supports :restart => true, :status => true, :reload => true
  action [:enable, :start]
end

```
Is this something that can be integrated into a future release?
I am not submitting a pull request as i am not familiar with chef-spec and am not able to write the specs to test this.
Postgres really doesn't like it if you change its data directory before stopping the service as it puts its PID file there and then looks at the postgres conf for where it should be.

Also it looks like the `execute 'Set locale and Create cluster' do` block will never execute as the postgres packages them selves tend to create the /etc/postgres... directories? 

Please let me know if there is a better way of doing this.

My testing of this was done in vagrant with the ubuntu/trusty64 image using the following role: 
```
name "postgres_testing"
description "Vagrant testing postgres"
run_list "recipe[postgresql::server]",
  "recipe[postgresql::ruby]"
override_attributes "postgresql" => {
    "version" => "9.3",
    "server" => { "packages" => [ "postgresql", "libpq-dev", "postgresql-contrib-9.3" ] },
    "password" => {
      "postgres" => "PASSWORD"
    },
    "config" => {
    "data_directory" => "/srv/postgresql/9.3/main",
    "listen_addresses" => "*",
    "log_min_duration_statement" => "1000",
    "log_line_prefix" => "%t %d "
    },
    "pg_hba" => [
      {:type => 'local', :db => 'all', :user => 'postgres', :addr => nil, :method => 'ident'},
      {:type => 'local', :db => 'all', :user => 'all', :addr => nil, :method => 'ident'},
      {:type => 'host', :db => 'all', :user => 'all', :addr => '127.0.0.1/32', :method => 'md5'},
      {:type => 'host', :db => 'all', :user => 'all', :addr => '::1/128', :method => 'md5'}
    ]
  }

**Comment by wojnosystems** [comment]
I'm having the same issue with Debian Wheezy.

```
Recipe: postgresql::server_debian
  * service[postgresql] action enable (up to date)
  * service[postgresql] action start (up to date)
  * execute[Set locale and Create cluster] action run
    
    ================================================================================
    Error executing action `run` on resource 'execute[Set locale and Create cluster]'
    ================================================================================
    
    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of export LC_ALL=C; /usr/bin/pg_createcluster --start 9.1.16 main ----
    STDOUT: 
    STDERR: Use of uninitialized value $version in concatenation (.) or string at /usr/bin/pg_createcluster line 263.
    Use of uninitialized value $version in concatenation (.) or string at /usr/bin/pg_createcluster line 273.
    Use of uninitialized value $version in concatenation (.) or string at /usr/bin/pg_createcluster line 277.
    Error: invalid version:
    ---- End output of export LC_ALL=C; /usr/bin/pg_createcluster --start 9.1.16 main ----
    Ran export LC_ALL=C; /usr/bin/pg_createcluster --start 9.1.16 main returned 1
    
    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/postgresql/recipes/server_debian.rb
    
     34: execute 'Set locale and Create cluster' do
     35:   command 'export LC_ALL=C; /usr/bin/pg_createcluster --start ' + node['postgresql']['version'] + ' main'
     36:   action :run
     37:   not_if { ::File.directory?('/etc/postgresql/' + node['postgresql']['version'] + '/main') }
     38: end
    
    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/postgresql/recipes/server_debian.rb:34:in `from_file'
    
    execute("Set locale and Create cluster") do
      action [:run]
      retries 0
      retry_delay 2
      guard_interpreter :default
      command "export LC_ALL=C; /usr/bin/pg_createcluster --start 9.1.16 main"
      backup 5
      returns 0
      cookbook_name "postgresql"
      recipe_name "server_debian"
      not_if { #code block }
    end
    

Running handlers:
[2015-08-20T16:52:43-07:00] ERROR: Running exception handlers
Running handlers complete
[2015-08-20T16:52:43-07:00] ERROR: Exception handlers complete
[2015-08-20T16:52:43-07:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
Chef Client failed. 16 resources updated in 48.017305287 seconds
[2015-08-20T16:52:43-07:00] ERROR: execute[Set locale and Create cluster] (postgresql::server_debian line 34) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of export LC_ALL=C; /usr/bin/pg_createcluster --start 9.1.16 main ----
STDOUT: 
STDERR: Use of uninitialized value $version in concatenation (.) or string at /usr/bin/pg_createcluster line 263.
Use of uninitialized value $version in concatenation (.) or string at /usr/bin/pg_createcluster line 273.
Use of uninitialized value $version in concatenation (.) or string at /usr/bin/pg_createcluster line 277.
Error: invalid version:
---- End output of export LC_ALL=C; /usr/bin/pg_createcluster --start 9.1.16 main ----
Ran export LC_ALL=C; /usr/bin/pg_createcluster --start 9.1.16 main returned 1
[2015-08-20T16:52:43-07:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

It appears as though the cookbook is not passing the version configuration properly. Postgresql seems to require a purely numerical version number, but the cookbook is passing the major/minor/sub version number:
http://osdir.com/ml/debian-bugs-dist/2010-06/msg03615.html

However, inspecting the pg_createcluster configuration script, version is being regexed as:
```
($version) = $ARGV[0] =~ /^(\d+\.\d+)$/;
```

Which means that only the MAJOR and MINOR numbers are snarfable and the cookbook is passing all 3, so the script doesn't read the version number at all.

So for my version: 9.1.16, I'd want to pass: 
```
/usr/bin/pg_createcluster --start 9.1 main
```

But the cookbook is passing:
```
/usr/bin/pg_createcluster --start 9.1.16 main
```

To fix this, the cookbook needs to pass only the MAJOR and MINOR numbers.

So the "postgresql/recipes/server_debian.rb":34-38 file should read:
```
execute 'Set locale and Create cluster' do
  command 'export LC_ALL=C; /usr/bin/pg_createcluster --start ' + node['postgresql']['version'].split('.')[0..1].join('.') + ' main'
  action :run
  not_if { ::File.directory?('/etc/postgresql/' + node['postgresql']['version'] + '/main') }
end
```

Maybe I'll submit a pull request after work.

**Comment by wojnosystems** [comment]
Pull request created for this. See: https://github.com/hw-cookbooks/postgresql/pull/285

**Comment by andre1810** [comment]
+1

**Comment by slyness** [comment]
This will function correctly soon in 4.0.0. #314

**Comment by slyness** [comment]
Resolved as of release 4.0.0 - https://github.com/hw-cookbooks/postgresql/releases/tag/v4.0.0

**Comment by phcyso** [comment]
Re opened this as #355 as it was not resolved in 4.0.0

**Comment by sworisbreathing** [comment]
@slyness please refer to #355 and/or update the documentation to clarify how node attributes should be overridden. The explicit calls to `node.set` in #314 break wrapper cookbooks which declare attributes at `default` level.
